design-proposals/vhostuser/vhostuser-interface.md

```diff
@@ -0,0 +1,166 @@
+
+# Overview
+- User Space networking is a feature that involves packet path from the VM to the data plane running in userspace like OVS-DPDK etc, by-passing the kernel.
+- With VhostUser interface, the kubevirt VM created can utilize the dpdk virtio functionality and avoid using kernel path for the traffic when DPDK is involved, allowing usage of fast datapath and improve performance.
+
+## Motivation
+- Any kubevirt VM uses linux kernel interfaces to connect to its host and to reach out externally or to communicate within the cluster. The current kubevirt VMs can only support the kernel data path for which the traffic needs to be through the kernel of the node/compute hosting the VM.
+
+- As a result, when we have a host with DPDK dataplane running in Userspace, the traffic from kubevirt VM will go through the kernel and reach the dataplane running in userspace. This longer path slows down the traffic between the application and the dataplane when both are running in userspace.
+
+- In order to improve the traffic performance and make use of the dpdk, we need to use a different path that can by-pass the kernel and reach the dataplane from the user space directly. This so called fast path will run within the user space and will allow fast packet processing.
+
+## Goals
+- The kubevirt VM when created with DPDK features, should be able to use fast datpath if the host supports the DPDK dataplane allowing fast packet processing to improve performance of the traffic.
+
+## Non Goals
+- The Userspace CNI with the dataplane running in userspace like dpdk/vpp, should make sure proper handling of mounting and unmounting the volumes. These changes are CNI specific and might vary from CNI to CNI.
```

Comment:
I'm not sure what the first sentence means?

The second sentence sounds like the non goal is "Automatically supporting existing CNI plugins. It will be necessary to modify existing CNI plugins in order to support vhost-user-net.".

Reply:
The changes proposed are specific to kubevirt itself. The changes are generic. Since interface addition and deletion is handled by a CNI, CNI should handle the mount and unmount the volumes where the sockets can be created.

This is not achieved by the changes proposed, User should make sure that CNI is capable of doing this or need to make necessary changes to the CNI to use this feature. This is the reason it was mentioned in the Non Goals section.